src/workerd/api/streams/standard.c++

```
@@ -3032,41 +3057,47 @@ private:
+                        (jsg::Lock& js, kj::Maybe<jsg::Value> maybeException) mutable {
+                      KJ_IF_MAYBE(reader, tryGetAs<PumpToReader>(pumpToReader)) {
+                        auto& ioContext = reader->ioContext;
+                        ioContext.requireCurrentOrThrowJs();
+                        // Oh good, if we got here it means we're in the right IoContext and
+                        // the PumpToReader is still alive.
+                        KJ_IF_MAYBE(exception, maybeException) {
+                          reader->doError(js, exception->getHandle(js));
+                        }
+                        return reader->pumpLoop(js, ioContext, kj::mv(readable), kj::mv(pumpToReader));
+                      } else {
+                        // If we got here, we're in the right IoContext but the PumpToReader
+                        // has been destroyed. Let's cancel the readable as the last step.
+                        return readable->cancel(js, maybeException.map([&](jsg::Value& ex) {
+                          return ex.getHandle(js);
+                        }));
                       }
-                      return reader->pumpLoop(js, ioContext, kj::mv(readable), kj::mv(pumpToReader));
-                    } else {
-                      // If we got here, we're in the right IoContext but the PumpToReader
-                      // has been destroyed. Let's cancel the readable as the last step.
-                      return readable->cancel(js, maybeException.map([&](jsg::Value& ex) {
-                        return ex.getHandle(js);
-                      }));
-                    }
-                  }));
+                    }));
+                  }, [&](jsg::Value exception) {
+                    // Exceptions here should be rare, and fairly odd. We would get here,
+                    // for instance, if the call to .Then inside read().then(...) failed.
+                    return handleException(kj::mv(exception));
```

Catching the exception and turning it into a rejected promise seems redundant here, since we're already in a `.then()` callback. If we just let the exception throw out of the callback, the promise machinery will already capture it and reject the promise.